Major speed-up of RGD scoring

ptosco opened the pull request:
This PR is about improving the R-group decomposition scoring speed by caching already settled matches rather than recomputing scores for all matches every time `process()` is called as currently happens.

The motivation for this PR is that I ran into an internal dataset of 567 molecules that was taking >10 minutes to process.
I analyzed our current RGD permutation scoring code, and I realized that since both the `Greedy` and `GreedyChunks` methods prune the "settled" match permutations to keep only the single best scoring permutation, there is no need to score all matches again and again every time `process()` is called, as their score will not vary anymore even though a new batch of molecules is added.
Therefore, I put some of the existing scoring logic into a new `RGroupScorer` class and implemented some caching logic in this class.
Then I made sure that, with the caching in place, I was still getting exactly the same results as I was getting from the current `master` branch.

The speed improvements are quite impressive, the more so the higher the degree of symmetry in the core(s), as symmetry makes the scoring step more time-consuming. Here follow some timings:
* the previously mentioned internal dataset (567 molecules) that used to take 11m7.673s now is processed in 2m11.138s (>5-fold speed-up)
* a public 1,4-DHP ChEMBL dataset that I assembled for benchmarking this PR (1069 1,4-dihydropyridines) that used to take 0m12.853s now is processed in 0m6.479s (~2-fold speed-up)
* a public bicyclo[2.2.2]octane ChEMBL dataset that I assembled for benchmarking this PR (1388 molecules) that used to take 233m51.679s now is processed in 40m58.195s (>5-fold speed-up)

See below some detailed statistics for the above three test cases, including the check of equality of PR results with the `master` branch results.

The Python benchmarking script:
[benchmark_rgd_scoring.py.gz](https://github.com/rdkit/rdkit/files/7206329/benchmark_rgd_scoring.py.gz)

The two public ChEMBL benchmark datasets:
[ChEMBL_bridged.smi.gz](https://github.com/rdkit/rdkit/files/7206333/ChEMBL_bridged.smi.gz)
[ChEMBL_1_4_DHP.smi.gz](https://github.com/rdkit/rdkit/files/7206331/ChEMBL_1_4_DHP.smi.gz)

# Internal 567-molecule dataset

### `master` branch

```
$ module purge; module load rdkit_master
$ time ( python rgd_perf.py )

real    11m14.800s
user    11m7.673s
sys     0m3.114s

base = "master"
with open(f"{base}_time_per_mol.pkl", "rb") as hnd:
    time_per_mol = pickle.load(hnd)
with open(f"{base}_results.pkl", "rb") as hnd:
    results = pickle.load(hnd)
with open(f"{base}_rejected.pkl", "rb") as hnd:
    rejected = pickle.load(hnd)

fig, ax = plt.subplots()
fig.set_size_inches(20, 10)
ax.plot(*zip(*time_per_mol))
plt.show()
```
![image](https://user-images.githubusercontent.com/5244385/134205803-0453e850-d22c-471d-b11e-4f80fe7939b2.png)
```
collected_results["master"] = (results, rejected)
```

### PR

```
$ module purge; module load rdkit_improve_rgd_scoring_speed
$ time ( python rgd_perf.py )

real    2m16.798s
user    2m11.138s
sys     0m3.024s


base = "improve_rgd_scoring_speed"
with open(f"{base}_time_per_mol.pkl", "rb") as hnd:
    time_per_mol = pickle.load(hnd)
with open(f"{base}_results.pkl", "rb") as hnd:
    results = pickle.load(hnd)
with open(f"{base}_rejected.pkl", "rb") as hnd:
    rejected = pickle.load(hnd)
fig, ax = plt.subplots()
fig.set_size_inches(20, 10)
ax.plot(*zip(*time_per_mol))
plt.show()
```
![image](https://user-images.githubusercontent.com/5244385/134206101-c2c767bc-8b56-4768-9c63-c338d31dc56c.png)
```
collected_results["PR"] = (results, rejected)
collected_results["master"] == collected_results["PR"]
True
len(collected_results["PR"][0]["Core"])
567
```

# ChEMBL 1,4-DHP 1069-molecule dataset

### `master` branch

```
$ module purge; module load rdkit_master
$ time ( dataset_scaffold='N1C=CCC=C1' dataset_base=ChEMBL_1_4_DHP python benchmark_rgd_scoring.py )

real    0m23.998s
user    0m12.853s
sys     0m1.398s

base = "ChEMBL_1_4_DHP_master/ChEMBL_1_4_DHP"
with open(f"{base}_time_per_mol.pkl", "rb") as hnd:
    time_per_mol = pickle.load(hnd)
with open(f"{base}_results.pkl", "rb") as hnd:
    results = pickle.load(hnd)
with open(f"{base}_rejected.pkl", "rb") as hnd:
    rejected = pickle.load(hnd)

fig, ax = plt.subplots()
fig.set_size_inches(20, 10)
ax.plot(*zip(*time_per_mol))
plt.show()
```
![image](https://user-images.githubusercontent.com/5244385/134245560-b607e131-90ad-47fc-828d-665b343af36c.png)
```
collected_results["master"] = (results, rejected)
```

### PR

```
$ module purge; module load rdkit_improve_rgd_scoring_speed
$ time ( dataset_scaffold='N1C=CCC=C1' dataset_base=ChEMBL_1_4_DHP python benchmark_rgd_scoring.py )

real    0m13.632s
user    0m6.479s
sys     0m1.298s

base = "ChEMBL_1_4_DHP_PR/ChEMBL_1_4_DHP"
with open(f"{base}_time_per_mol.pkl", "rb") as hnd:
    time_per_mol = pickle.load(hnd)
with open(f"{base}_results.pkl", "rb") as hnd:
    results = pickle.load(hnd)
with open(f"{base}_rejected.pkl", "rb") as hnd:
    rejected = pickle.load(hnd)

fig, ax = plt.subplots()
fig.set_size_inches(20, 10)
ax.plot(*zip(*time_per_mol))
plt.show()
```
![image](https://user-images.githubusercontent.com/5244385/134245791-892d3142-9e08-4e9d-b465-203a12342743.png)
```
collected_results["PR"] = (results, rejected)
collected_results["master"] == collected_results["PR"]
len(collected_results["master"][0]["Core"])
1069
```

# ChEMBL bicyclo[2.2.2]octane 1388-molecule dataset

### `master` branch

```
$ module purge; module load rdkit_master
$ time ( dataset_scaffold='C1CC2CCC1CC2' dataset_base=ChEMBL_bridged python benchmark_rgd_scoring.py )

real    234m43.518s
user    233m51.679s
sys     0m17.037s

base = "master/ChEMBL_bridged"
with open(f"{base}_time_per_mol.pkl", "rb") as hnd:
    time_per_mol = pickle.load(hnd)
with open(f"{base}_results.pkl", "rb") as hnd:
    results = pickle.load(hnd)
with open(f"{base}_rejected.pkl", "rb") as hnd:
    rejected = pickle.load(hnd)

fig, ax = plt.subplots()
fig.set_size_inches(20, 10)
ax.plot(*zip(*time_per_mol))
plt.show()
```
![image](https://user-images.githubusercontent.com/5244385/134246140-ec789979-fe9a-4132-a3c5-fb5700984f3e.png)
```
collected_results["master"] = (results, rejected)
```

### PR

```
$ module purge; module load rdkit_improve_rgd_scoring_speed
$ time ( dataset_scaffold='C1CC2CCC1CC2' dataset_base=ChEMBL_bridged python benchmark_rgd_scoring.py )

real    42m15.375s
user    40m58.195s
sys     1m6.836s

base = "ChEMBL_bridged_PR/ChEMBL_bridged"
with open(f"{base}_time_per_mol.pkl", "rb") as hnd:
    time_per_mol = pickle.load(hnd)
with open(f"{base}_results.pkl", "rb") as hnd:
    results = pickle.load(hnd)
with open(f"{base}_rejected.pkl", "rb") as hnd:
    rejected = pickle.load(hnd)

fig, ax = plt.subplots()
fig.set_size_inches(20, 10)
ax.plot(*zip(*time_per_mol))
plt.show()
```
![image](https://user-images.githubusercontent.com/5244385/134246316-9107f222-468e-4e49-9976-1ef5fb297399.png)
```
collected_results["PR"] = (results, rejected)
collected_results["master"] == collected_results["PR"]
True
len(collected_results["`master` branch"][0]["Core"])
1388
```

ptosco commented:
The failure in the `Ubuntu_18_04_x64` build is unrelated to this PR (I think I have occasionally seen this before):
```
194/206 Test #194: pySimDivPickers ........................***Failed    0.40 sec
[22:11:49] the useCache argument is deprecated and ignored
...[22:11:49] the useCache argument is deprecated and ignored
[22:11:49] the useCache argument is deprecated and ignored
.[22:11:49] the useCache argument is deprecated and ignored
......F
======================================================================
FAIL: testNonUniqueCrash (__main__.TestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vsts/work/1/s/Code/SimDivPickers/Wrap/testPickers.py", line 139, in testNonUniqueCrash
    self.assertTrue(tuple(mm2) != tuple(mm1)) or (tuple(mm3) != tuple(mm1))
AssertionError: False is not true

----------------------------------------------------------------------
Ran 11 tests in 0.181s

FAILED (failures=1)
```

greglandrum commented:
> The failure in the `Ubuntu_18_04_x64` build is unrelated to this PR (I think I have occasionally seen this before):

yeah, that's definitely not you

bp-kelley commented:
@ptosco thanks for doing this, I've had it on my roadmap for a while, but hadn't got around to it.

This will take a bit to review, but I'll start it after I do the reaction review.

greglandrum commented on Code/GraphMol/RGroupDecomposition/RGroupScore.h:
```diff
@@ -67,10 +68,81 @@ struct CartesianProduct {
+      // heavyCounts is a vector which has the same size of labels
+      // for each label we add an increment if a molecule
+      // bears an R-group at that label
+      size_t offset = matches.size() - permutation.size();
```
you need a PRECONDITION here on the sizes

greglandrum reviewed (changes requested):
This wasn't looking at design, algorithm, or correctness (the tests take care of that last one).
I was mainly looking at the code itself.

ptosco commented on Code/GraphMol/RGroupDecomposition/RGroupDecomp.cpp:
```diff
@@ -369,28 +369,32 @@ int RGroupDecomposition::add(const ROMol &inmol) {
     return -2;
   }
 
+  const bool prune = true;
+  if (prune) {
```
There is a rationale here. The current code always passes `prune = true` to `process(0`. To make code more readable, I define a `const` as we do elsewhere. Regarding the check: I know that the check at the moment is just a waste of CPU cycles, but I would like that the code still behaves if someone for some reason decides to set `prune` to `false` in his own fork of the code rather then just assuming that prune is always `true`.

greglandrum replied:
maybe just a quick comment to that effect?
```suggestion
  const bool prune = true;
  // in case the value ends up being changed in a future version of the code:
  if (prune) {
```

ptosco commented on Code/GraphMol/RGroupDecomposition/RGroupDecompData.h:
```diff
@@ -121,13 +123,15 @@ struct RGroupDecompData {
   }
 
   void prune() {  // prune all but the current "best" permutation of matches
+    size_t offset = matches.size() - permutation.size();
```
We are. However, I can add a `CHECK_INVARIANT`, one more check won't make a big difference.

greglandrum replied:
thanks

greglandrum commented:
I forgot to comment on this in the original review: this kind of speedup is awesome to get and the way you're doing it seems really sensible to me.

ptosco commented:
@greglandrum I think I have addressed all your comments.

bp-kelley commented on Code/GraphMol/RGroupDecomposition/RGroupDecompData.h:
```diff
@@ -27,6 +27,8 @@ struct RGroupDecompData {
   std::map<std::string, int> newCores;  // new "cores" found along the way
   int newCoreLabel = EMPTY_CORE_LABEL;
+  size_t permutationProduct = 1;
+  size_t previousMatchSize = 0;
```
Could we comment these variables, at least their intent.

bp-kelley commented on Code/GraphMol/RGroupDecomposition/RGroupScore.h:
```diff
@@ -67,10 +68,82 @@ struct CartesianProduct {
+    std::vector<size_t> permutation;
+  };
+  double d_bestScore = 0.0;
+  State d_current;
```
Could you comment the logic of this caching.  current and initial are understandable, but saved and store seem redundant.

Also, "State" is probably not descriptive enough.

bp-kelley commented on Code/GraphMol/RGroupDecomposition/RGroupScore.h:
```diff
@@ -67,10 +68,82 @@ struct CartesianProduct {
-    const std::vector<size_t> &permutation,
-    const std::vector<std::vector<RGroupMatch>> &matches,
-    const std::set<int> &labels);
+class RDKIT_RGROUPDECOMPOSITION_EXPORT RGroupScorer {
```
Now that the scorer is an independent class, we might want to have some tests.

bp-kelley replied:
I also think that not that this class has new functionality, it should be documented.

bp-kelley commented on Code/GraphMol/RGroupDecomposition/RGroupScore.h:
```diff
@@ -67,10 +68,82 @@ struct CartesianProduct {
+      size_t i = 0;
+      for (int label : orderedLabels) {
+        for (size_t m = 0; m < permutation.size(); ++m) {  // for each molecule
+          // for each molecule, check if we add an R-group at this negative
```
For my sanity, it would be good to reiterate that negative labels are newly found and not user specified.

bp-kelley commented:
@ptosco @greglandrum for some reason, I'm not seeing the discussion/resolution on your comments above so the logic train is a bit confusing.

bp-kelley commented on Code/GraphMol/RGroupDecomposition/RGroupDecomp.cpp:
```diff
@@ -370,6 +370,7 @@ int RGroupDecomposition::add(const ROMol &inmol) {
   }
 
   const bool prune = true;
+  // in case the value ends up being changed in a future version of the code:
```
I really don't think this is useful and is confusing when reviewing the code.  I don't mind the const bool prune, but the if statement below screams to be gone.  I vote for getting rid of it and making prune the default.  If we want to be able to change it in the future, make it a member variable, but this code is complex enough removing dead code is a good thing :)

bp-kelley commented on Code/GraphMol/RGroupDecomposition/RGroupGa.cpp:
```diff
@@ -294,7 +294,7 @@ GaResult RGroupGa::run(int runNumber) {
   BOOST_LOG(rdInfoLog) << "Run " << runNumber << " Execution "
                        << timeInfo(startTime) << std::endl;
-  GaResult result(best->getFitness(), permutations);
+  GaResult result{best->getFitness(), permutations};
```
I find this change confusing as well, I tend to use {} for initializer lists which this is clearly not.  But that is my style, however I think we should be consistent.

bp-kelley commented on Code/GraphMol/RGroupDecomposition/RGroupDecomp.cpp:
```diff
@@ -369,28 +369,33 @@ int RGroupDecomposition::add(const ROMol &inmol) {
     if (N * potentialMatches.size() > 100000) {
-      data->permutation = std::vector<size_t>(data->matches.size(), 0);
-      data->process(true);
+      data->permutationProduct = N;
```
Have we tried an example where this is the case?  I know the older behavior, but not the new one.  we didn't have a real test for this before except for running it on larger datasets,

bp-kelley commented on Code/GraphMol/RGroupDecomposition/RGroupGa.h:
```diff
@@ -103,7 +102,7 @@ struct GaResult {
 class RDKIT_RGROUPDECOMPOSITION_EXPORT RGroupGa : public GaBase {
  public:
-  RGroupGa(const RGroupDecompData& rGroupData,
+  RGroupGa(RGroupDecompData& rGroupData,
```
I believe we could keep this const by making the RGroupScorer mutable.  Since the scorer is just caching and not changing results, this its kosher and more explicit in what it is doing I think.

bp-kelley replied:
(I haven't double checked this, so I could be wrong). But I like int better than changing the API.

bp-kelley commented on Code/GraphMol/RGroupDecomposition/RGroupScore.cpp:
```diff
@@ -33,38 +37,36 @@ double matchScore(const std::vector<size_t> &permutation,
 
   // What is the largest rgroup count at any label
-  int N = 0;
+  restoreInitialState();
```
stopping here for now ...

greglandrum commented on Code/GraphMol/RGroupDecomposition/RGroupGa.h:
```diff
@@ -120,7 +119,7 @@ class RDKIT_RGROUPDECOMPOSITION_EXPORT RGroupGa : public GaBase {
   shared_ptr<RGroupDecompositionChromosome> createChromosome();
 
-  const RGroupDecompData& getRGroupData() const { return rGroupData; }
+  RGroupDecompData& getRGroupData() const { return rGroupData; }
```
@ptosco : this was marked as resolved but there's neither a reply to the question nor a code change

greglandrum commented on Code/GraphMol/RGroupDecomposition/RGroupGa.h:
```diff
@@ -130,7 +129,7 @@ class RDKIT_RGROUPDECOMPOSITION_EXPORT RGroupGa : public GaBase {
   RGroupGa(const RGroupGa& other) = delete;
   RGroupGa& operator=(const RGroupGa& other) = delete;
-  const RGroupDecompData& rGroupData;
+  RGroupDecompData& rGroupData;
```
@ptosco : this was marked as resolved but there's neither a reply to the question nor a code change

ptosco commented:
@bp-kelley Hi Brian, I think I have addressed all your comments except unit tests for `RGroupScorer`.
I have duly checked that results haven’t changed compared to `master` (see above), and that should be sufficient.
Regarding testing the case where permutations exceed 1e6, yes, I have tested that case as part of the tests described above in the PR. Some of those datasets actually do incur into that situation and are therefore processed before `chunkSize` is reached and, yes, I have verified that the caching mechanism indeed works also in that case.

greglandrum reviewed (approved):
LGTM

bp-kelley commented:
@ptosco Nice, I have one more trick up my sleeve for a speed improvement and this foundation will help a ton I think.
